### PR TITLE
feature: add thread_pool

### PR DESCRIPTION
**Pull request by code0xff**

### CODE REVIEW

**Review by lazka33 — COMMENTED** [review]
LGTM

File: src/noir/common/thread_pool.cpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+void named_thread_pool::stop() {
+  _ioc_work.reset();
+  _ioc.stop();
+  _thread_pool.join();
+  _thread_pool.stop();
+}
+
+}
```
Comment (xedonman) [inline src/noir/common/thread_pool.cpp]: need new line

File: src/noir/common/thread_pool.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+// async on thread_pool and return future
+template<typename F>
+auto async_thread_pool(boost::asio::io_context& thread_pool, F&& f) {
+  auto task = std::make_shared<std::packaged_task<decltype(f())()>>(std::forward<F>(f));
+  boost::asio::post(thread_pool, [task]() { (*task)(); });
+  return task->get_future();
+}
+
+}
```
Comment (xedonman) [inline src/noir/common/thread_pool.h]: need new line

**Review by xedonman — COMMENTED** [review]
lgtm, plz check comment for new line

**Review by conr2d — APPROVED** [review]

**Pull request merged**